# How to handle 10X VDJ sequencing results
For our project, we would like to test the most prominent TIL TCRs discovered
against the cell line of interest and see if it is really reactive against the
corresponding cell line (`MC38B` or `LLC-AF91`). This means that, seeing
enrichments based on the `CDR3` sequence is great but what we need is the
complete TCR sequence so that we can get it cloned into a mammalian expression
vector. This document outlines the step to go from the 10X VDJ profiling
results to ordering constructs.

## 1. Extract the 10X sequencing results for the TCR
Open the loupe browser and pick the TCR chain you would like to work on:

![](https://cl.ly/901d5d23daca/%255Be012cc6333ece07d000d8f836a5e3f3e%255D_Image%2525202019-08-06%252520at%2525202.35.22%252520PM.png)

Export the results into a FASTA file:

![](https://cl.ly/a9365cf439b9/%255B8a2472dd688d95f1de09afe9fe6650d4%255D_Image%2525202019-08-06%252520at%2525202.37.10%252520PM.png)

which should look something like this:

```
>clonotype1_concat_ref_1
ATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGATGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCTGGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAATGGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGTGTACTTCTGTGCTGTGAGCGTGGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

>clonotype1_consensus_1
TGGGGGTGCATTTCCTGTTGCCTCTGCAACATGCTGCAGAGCATCTACCAAGGGCAATATCTGTGTCATATAGTGTCACTACCATGGCTGTGGTGGTCATTCACACTCAGCCTGAGACAATTGCATCTGAAGTCTGGGGTACACTGCATAACCTGGTTTCCTAGGGAGTATTAACAAGCCATTCACTTTCACTAAATGCTAAAACCAATATATACCAAATTAGGAAAAGATACAATGTGCTTTGTGGGAGTTTTACTTTGTAAAACAGGAGCATAGCAGGAGGCAGGCTGTGGACCCTCAAGACATTTGTTTCCTCCTATGGCCTCCAGAGGGCGCTGCTGCACTCACTAGTGTTCCTTCTTCCTTTGAAGATCTCCACCTGCTCACATCTCATCCTTCTTGCAGCTGGCTGAGTATATAGCAAAGGCAAGTGACAAGAATCAGATATTCAAAACCTTCCAGGCTCGTGTTTGCAGCGTCAGTTTTTATGTGAGCGGGAGACTTCTCCAGCAGACTCTTCCCTACACCTGCTTTCAAGCTCTTGCTTTCCTCAGCCTCTTTCCTGCCTGTCCTGTTCCAGAGTTCCTCCACAACAGAGCTGCAGCCTTCCCAAGGCTCAGCCATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGATGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCTGGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAATGGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGTGTACTTCTGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

>AAACCTGCATGGTAGG-1_contig_1
GGGCACATCTCATCCTTCTTGCAGCTGGCTGAGTATATAGCAAAGGCAAGTGACAAGAATCAGATATTCAAAACCTTCCAGGCTCGTGTTTGCAGCGTCAGTTTTTATGTGAGCGGGAGACTTCTCCAGCAGACTCTTCCCTACACCTGCTTTCAAGCTCTTGCTTTCCTCAGCCTCTTTCCTGCCTGTCCTGTTCCAGAGTTCCTCCACAACAGAGCTGCAGCCTTCCCAAGGCTCAGCCATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGATGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCTGGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAATGGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGTGTACTTCTGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

...
```

## 2. Copy/paste the 5' end starting from the `ATG` codon
Note that the first two sequences are the ones we are interested in.
The `*_ref_*` represents the reference genome and it doesn't contain the
unique `CDR3` region that is very crucial for us. So make sure to copy the
sequence from the `*_consensus_*` one. Often times, consensus leaks into the
5' UTR region, which we don't want to clone. So make sure you identify the
coding region right:

```
>clonotype1_concat_ref_1
[ATGCTCCTGGCACTCCTCCCAGTGCTGGGG] ATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCC
GATGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTT
CTGGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGA
ATGGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCT
GTGTACTTCTGTGCTGTGAGCGTGGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGA
CATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACT
TTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCT
ATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

>clonotype1_consensus_1
TGGGGGTGCATTTCCTGTTGCCTCTGCAACATGCTGCAGAGCATCTACCAAGGGCAATATCTGTGTCATATAGTGTCACT
ACCATGGCTGTGGTGGTCATTCACACTCAGCCTGAGACAATTGCATCTGAAGTCTGGGGTACACTGCATAACCTGGTTTC
CTAGGGAGTATTAACAAGCCATTCACTTTCACTAAATGCTAAAACCAATATATACCAAATTAGGAAAAGATACAATGTGC
TTTGTGGGAGTTTTACTTTGTAAAACAGGAGCATAGCAGGAGGCAGGCTGTGGACCCTCAAGACATTTGTTTCCTCCTAT
GGCCTCCAGAGGGCGCTGCTGCACTCACTAGTGTTCCTTCTTCCTTTGAAGATCTCCACCTGCTCACATCTCATCCTTCT
TGCAGCTGGCTGAGTATATAGCAAAGGCAAGTGACAAGAATCAGATATTCAAAACCTTCCAGGCTCGTGTTTGCAGCGTC
AGTTTTTATGTGAGCGGGAGACTTCTCCAGCAGACTCTTCCCTACACCTGCTTTCAAGCTCTTGCTTTCCTCAGCCTCTT
TCCTGCCTGTCCTGTTCCAGAGTTCCTCCACAACAGAGCTGCAGCCTTCCCAAGGCTCAGCC [ATGCTCCTGGCACTCCT
CCCAGTGCTGGGG] ATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGATGCTCGCGTCACTGTC
TCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCTGGTATGTCCAGTACCC
GCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAATGGCTTTGAGGCTGAGT
TCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGTGTACTTCTGTGCTGTG
AGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACATCCAGAACCCAGAACC
TGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATG
TGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAAT
GGGGCCATTGCCTGGAGCAACCAG
```

so the sequence we are interested in is highlighted within `[]`:
```
>clonotype1_concat_ref_1
ATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGA
TGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCT
GGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAAT
GGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGT
GTACTTCTGTGCTGTGAGCGTGGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACA
TCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTT
GACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTAT
GGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

>clonotype1_consensus_1
TGGGGGTGCATTTCCTGTTGCCTCTGCAACATGCTGCAGAGCATCTACCAAGGGCAATATCTGTGTCATATAGTGTCACT
ACCATGGCTGTGGTGGTCATTCACACTCAGCCTGAGACAATTGCATCTGAAGTCTGGGGTACACTGCATAACCTGGTTTC
CTAGGGAGTATTAACAAGCCATTCACTTTCACTAAATGCTAAAACCAATATATACCAAATTAGGAAAAGATACAATGTGC
TTTGTGGGAGTTTTACTTTGTAAAACAGGAGCATAGCAGGAGGCAGGCTGTGGACCCTCAAGACATTTGTTTCCTCCTAT
GGCCTCCAGAGGGCGCTGCTGCACTCACTAGTGTTCCTTCTTCCTTTGAAGATCTCCACCTGCTCACATCTCATCCTTCT
TGCAGCTGGCTGAGTATATAGCAAAGGCAAGTGACAAGAATCAGATATTCAAAACCTTCCAGGCTCGTGTTTGCAGCGTC
AGTTTTTATGTGAGCGGGAGACTTCTCCAGCAGACTCTTCCCTACACCTGCTTTCAAGCTCTTGCTTTCCTCAGCCTCTT
TCCTGCCTGTCCTGTTCCAGAGTTCCTCCACAACAGAGCTGCAGCCTTCCCAAGGCTCAGCC [ATGCTCCTGGCACTCC
TCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGATGCTCGCGTCACTGTC
TCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCTGGTATGTCCAGTACCC
GCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAATGGCTTTGAGGCTGAGT
TCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGTGTACTTCTGTGCTGTG
AGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACATCCAGAACCCAGAACC
TGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATG
TGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAAT
GGGGCCATTGCCTGGAGCAACCAG]
```

If the `*_ref_*` sequence also leaks into the 5' region, then use the Loupe
Browser to identify where the `V` segment starts:

![](https://cl.ly/1648b6b9a4bf/%255B6bc26cbe30cf9e071c18e445ea3615eb%255D_Image%2525202019-08-06%252520at%2525202.49.28%252520PM.png)

Now, it is a good idea to keep track of the `CDR3` region and the type of chain
we are working with:

![](https://cl.ly/5b35d530bb79/%255B6b1729baa71422d121131f0e0df10933%255D_Image%2525202019-08-06%252520at%2525202.48.29%252520PM.png)

A good way to do this is to encode these in the sequence description:

```
> Sample32_TRA_CAVSDDSNYQLIW_TGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGG
ATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGA
TGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCT
GGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAAT
GGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGT
GTACTTCTGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACA
TCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTT
GACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTAT
GGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG
```

## 3. Complete the constant region sequence
Often times, the 10X VDJ sequencing doesn't go too much into the constant
region, so both the `*_ref_*` and the `*_consensus_*` sequences will lack the
complete constant sequence. We have to manually fill this in. In this example,
our samples were of the lab stain background, so we already know the sequences
of the constant regions for both chains:

- Mouse TRAC (alpha): [mouse.TRAC.dna.fasta](https://github.com/hammerlab/tcr-transfer/blob/master/sequences/mouse.TRAC.dna.fasta)
- Mouse TRBC (beta): [mouse.TRBC.dna.fasta](https://github.com/hammerlab/tcr-transfer/blob/master/sequences/mouse.TRBC.dna.fasta)

Given the example sequence we have copied was coming from the alpha chain,
we would like to complete the sequence with the alpha constant region. This is
usually a matter of finding a good overlap and filling in the 3' portion:

```
> Sample32_TRA_CAVSDDSNYQLIW_TGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGG
ATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGA
TGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCT
GGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAAT
GGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGT
GTACTTCTGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAG [A
CATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGG] ACAGCACCCTCTGCCTGTTCACCGA
CTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAG
CTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAG

> Mouse_TRAC
[ACATCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGG] ACAGCACCCTCTGCCTGTTCAC
CGACTTTGACTCCCAAATCAATGTGCCGAAAACCATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGA
AAGCTATGGATTCCAAGAGCAATGGGGCCATTGCCTGGAGCAACCAGACAAGCTTCACCTGCCAAGATATCTTCAAAGAG
ACCAACGCCACCTACCCCAGTTCAGACGTTCCCTGTGATGCCACGTTGACTGAGAAAAGCTTTGAAACAGATATGAACCT
AAACTTTCAAAACCTGTCAGTTATGGGACTCCGAATCCTCCTGCTGAAAGTAGCCGGATTTAACCTGCTCATGACGCTGA
GGCTGTGGTCCTCC

```

so the final TCR sequence will be:

```
> Sample32_TRA_CAVSDDSNYQLIW_TGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGG
ATGCTCCTGGCACTCCTCCCAGTGCTGGGGATACACTTTGTCCTGAGAGATGCCCAAGCTCAGTCAGTGACACAGCCCGA
TGCTCGCGTCACTGTCTCTGAAGGAGCCTCTCTGCAGCTGAGATGCAAGTATTCCTACTCTGGGACGCCTTATCTGTTCT
GGTATGTCCAGTACCCGCGGCAGGGGCTGCAGCTGCTCCTCAAGTACTATTCCGGAGACCCAGTGGTTCAAGGAGTGAAT
GGCTTTGAGGCTGAGTTCATCAAGAGTAACTCTTCCTTCCACCTGCGGAAAGCCTCTGTGCACTGGAGCGACTCGGCTGT
GTACTTCTGTGCTGTGAGCGACGATAGCAACTATCAGTTGATCTGGGGCTCTGGGACCAAGCTAATTATAAAGCCAGACA
TCCAGAACCCAGAACCTGCTGTGTACCAGTTAAAAGATCCTCGGTCTCAGGACATCCAGAACCCAGAACCTGCTGTGTAC
CAGTTAAAAGATCCTCGGTCTCAGGACAGCACCCTCTGCCTGTTCACCGACTTTGACTCCCAAATCAATGTGCCGAAAAC
CATGGAATCTGGAACGTTCATCACTGACAAAACTGTGCTGGACATGAAAGCTATGGATTCCAAGAGCAATGGGGCCATTG
CCTGGAGCAACCAGACAAGCTTCACCTGCCAAGATATCTTCAAAGAGACCAACGCCACCTACCCCAGTTCAGACGTTCCC
TGTGATGCCACGTTGACTGAGAAAAGCTTTGAAACAGATATGAACCTAAACTTTCAAAACCTGTCAGTTATGGGACTCCG
AATCCTCCTGCTGAAAGTAGCCGGATTTAACCTGCTCATGACGCTGAGGCTGTGGTCCTCC
```

## 4. Sanity checks
- Make sure the DNA sequence starts with `ATG`;
- Make sure the DNA sequence contains the DNA `CDR3` sequence;
- Make sure the translated sequence (e.g. via
  [ExPASy's translate tool](https://web.expasy.org/translate/)
  contains the aminoacid `CDR3` sequence;
- Make sure the translated sequence starts with the same `*_consensus_*` that
  10X predicted. Use `Export > Aminoacid sequence` dialog for this.

If you'd like to automate these sanity checks, here is one way of doing it
via BioPython: [Check and translate TCR DNA sequences.ipynb](https://github.com/hammerlab/tcr-transfer/blob/master/analyses/Check%20and%20translate%20TCR%20DNA%20sequences.ipynb).

## 5. Optional customization for cloning
- Add a stop codon to the very end;
- Add restriction enzyme sites that do not exist within the TCR sequence
  (see [Pick the best enzymes for cloning all TCRs into the plasmid.ipynb](https://github.com/hammerlab/tcr-transfer/blob/master/analyses/Pick%20the%20best%20enzymes%20for%20cloning%20all%20TCRs%20into%20the%20plasmid.ipynb)) for example to who to best automate this.
